experimental/async_impl.hpp

```diff
@@ -0,0 +1,565 @@
+template <typename _IsSync = ::std::true_type,
+          __par_backend_hetero::access_mode __acc_mode1 = __par_backend_hetero::access_mode::read,
+          __par_backend_hetero::access_mode __acc_mode2 = __par_backend_hetero::access_mode::write,
+          typename _ExecutionPolicy, typename _ForwardIterator1, typename _ForwardIterator2, typename _Function>
+oneapi::dpl::__internal::__enable_if_async_execution_policy<
+    _ExecutionPolicy, oneapi::dpl::__internal::__future<_ForwardIterator2>>
+__pattern_walk2_async(_ExecutionPolicy&& __exec, _ForwardIterator1 __first1, _ForwardIterator1 __last1,
+                      _ForwardIterator2 __first2, _Function __f, /*vector=*/::std::true_type,
+                      /*parallel=*/::std::true_type)
+{
+    auto __n = __last1 - __first1;
+    if (__n <= 0)
+        return oneapi::dpl::__internal::__future<_ForwardIterator2>( sycl::event{}, __first2);
+
+    auto __keep1 = oneapi::dpl::__ranges::__get_sycl_range<__acc_mode1, _ForwardIterator1>();
+    auto __buf1 = __keep1(__first1, __last1);
+
+    auto __keep2 = oneapi::dpl::__ranges::__get_sycl_range<__acc_mode2, _ForwardIterator2>();
+    auto __buf2 = __keep2(__first2, __first2 + __n);
+
+    auto __future_obj = oneapi::dpl::__par_backend_hetero::__parallel_for(
+        ::std::forward<_ExecutionPolicy>(__exec), unseq_backend::walk_n<_ExecutionPolicy, _Function>{__f}, __n,
+        __buf1.all_view(), __buf2.all_view());
+    oneapi::dpl::__internal::__invoke_if(_IsSync(), [&__future_obj]() { __future_obj.wait(); });
+
+    return oneapi::dpl::__internal::__future<_ForwardIterator2>(__future_obj, __first2 + __n);
+}
+
+template <typename _ExecutionPolicy, typename _ForwardIterator1, typename _ForwardIterator2, typename _Brick>
+oneapi::dpl::__internal::__enable_if_async_execution_policy<
+    _ExecutionPolicy, oneapi::dpl::__internal::__future<_ForwardIterator2>>
+__pattern_walk2_brick_async(_ExecutionPolicy&& __exec, _ForwardIterator1 __first1, _ForwardIterator1 __last1,
+                            _ForwardIterator2 __first2, _Brick __brick, /*parallel*/ ::std::true_type)
+{
+    return __pattern_walk2_async(
+        __par_backend_hetero::make_wrapped_policy<__walk2_brick_wrapper>(::std::forward<_ExecutionPolicy>(__exec)),
+        __first1, __last1, __first2, __brick,
+        /*vector=*/::std::true_type{}, /*parallel*/ ::std::true_type{});
+}
+
+//------------------------------------------------------------------------
+// transform_reduce (with unary and binary functions)
+//------------------------------------------------------------------------
+
+template <typename _ExecutionPolicy, typename _ForwardIterator, typename _Tp, typename _BinaryOperation,
+          typename _UnaryOperation>
+oneapi::dpl::__internal::__enable_if_async_execution_policy<
+    _ExecutionPolicy,
+    //_Tp>
+    oneapi::dpl::__internal::__future<_Tp>>
+__pattern_transform_reduce_async(_ExecutionPolicy&& __exec, _ForwardIterator __first, _ForwardIterator __last,
```

Comment:
This should be placed in `numeric_impl_hetero.h`

Reply:
we agreed offline to keep it in separate files for now.